ee/modules/110-istio/hooks/ee/alliance_metadata_merge.go

```diff
@@ -126,6 +132,180 @@ func applyMulticlusterMergeFilter(obj *unstructured.Unstructured) (go_hook.Filte
+	if err != nil {
+		return TokenValidationResult{
+			IsValid: false,
+			Error:   fmt.Sprintf("failed to parse token: %v", err),
+		}
+	}
+
+	// Check expiration time
+	payload := token.UnsafePayloadWithoutVerification()
+
+	// Parse the payload
+	var claims map[string]interface{}
+	if err := json.Unmarshal(payload, &claims); err != nil {
+		return TokenValidationResult{
+			IsValid: false,
+			Error:   fmt.Sprintf("failed to unmarshal claims: %v", err),
+		}
+	}
+
+	// Check if exp claim exists
+	expClaim, exists := claims["exp"]
+	if !exists {
+		return TokenValidationResult{
+			IsValid: false,
+			Error:   "token has no expiration claim",
+		}
+	}
+
+	var expTime time.Time
+	switch exp := expClaim.(type) {
+	case float64:
+		expTime = time.Unix(int64(exp), 0)
+	case int64:
+		expTime = time.Unix(exp, 0)
+	default:
+		return TokenValidationResult{
+			IsValid: false,
+			Error:   "invalid expiration claim format",
+		}
+	}
+
+	now := time.Now()
+	isExpired := now.After(expTime)
+
+	return TokenValidationResult{
+		IsValid:   !isExpired,
+		IsExpired: isExpired,
+		ExpiresAt: expTime,
+	}
+}
+
+func applyIstioRemoteSecretFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	secret := &v1.Secret{}
+	err := sdk.FromUnstructured(obj, secret)
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert k8s secret to struct: %v", err)
+	}
+
+	secretName := secret.GetName()
+	if !strings.HasPrefix(secretName, "istio-remote-secret-") {
+		return nil, fmt.Errorf("secret %s is not an istio remote secret", secretName)
+	}
+	clusterName := strings.TrimPrefix(secretName, "istio-remote-secret-")
+
+	// Get the base64-encoded kubeconfig from the field named after the cluster
+	secData, exists := secret.Data[clusterName]
+	if !exists {
+		return nil, fmt.Errorf("secret %s does not contain '%s' field", secretName, clusterName)
+	}
+
+	// Handle the kubeconfig data - it might be base64-encoded or raw YAML
+	secDataStr := string(secData)
+	var kubeconfigBytes []byte
+
+	// Check if the data looks like base64 (contains only base64 characters)
+	if isBase64String(secDataStr) {
+		var err error
+		kubeconfigBytes, err = base64.StdEncoding.DecodeString(secDataStr)
+		if err != nil {
+			return nil, fmt.Errorf("cannot decode base64 kubeconfig from secret %s: %v", secretName, err)
+		}
+	} else {
+		kubeconfigBytes = secData
+	}
+
+	// Extract token from kubeconfig directly
+	var kubeconfig Kubeconfig
+	err = yaml.Unmarshal(kubeconfigBytes, &kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal kubeconfig from secret %s: %v", secretName, err)
+	}
+
+	// Look for the first user with a token
+	var token string
+	for _, user := range kubeconfig.Users {
+		if user.User.Token != "" {
+			token = user.User.Token
+			break
+		}
+	}
+
+	if token == "" {
+		return nil, fmt.Errorf("token not found in kubeconfig from secret %s", secretName)
+	}
+
+	return IstioRemoteSecretToken{
+		Name:  secretName,
```

Comment:
```suggestion
		MultiClusterName:  annotations['networking.istio.io/cluster'],
```

Reply:
Thank you, i change logic for use annotations from secrets with multicluster kubeconfig.